File: multithread_cpp/prime.cpp
#include<thread>
#include<stdio.h>

using namespace std;

#define STEP  10000

/**
 * g++ -std=c++11 -I/usr/local/lib/gcc49/include/c++/ -Wl,-rpath=/usr/local/lib/gcc49 -pthread prime.cpp
 */

bool isPrime(int number)
{
	for(int n = 2; n <= number / 2; n++)
	{
		if(number % n == 0) return false;
	}
	return true;
}

void work(bool list[], int start, int end, int offset)
{
	for(int number = start; number < end; number+=4) list[number-offset] = isPrime(number);
}

int main(int argc, char *args[])
{
	int limit = 0;
	if(argc == 2) limit = atoi(args[1]);
	if(limit < 1) limit = 10000;

	clock_t start = clock();
	int count = 0;
	for(int number = 2; ; number += STEP)
	{
		bool list[STEP];
		std::thread t0(work,list,number+0,number+STEP,number);
		std::thread t1(work,list,number+1,number+STEP,number);
		std::thread t2(work,list,number+2,number+STEP,number);
		std::thread t3(work,list,number+3,number+STEP,number);
		std::this_thread::sleep_for(std::chrono::microseconds(1));
		t0.join();
		t1.join();
		t2.join();
		t3.join();

		for(int s = 0; s < STEP; s++)
		{
			if(list[s]) count++;
			if(count == limit)
			{
				printf("%dth prime is %d\n", limit, number+s);
				goto end;
			}
		}
	}
	end:
	printf("[mcpp] duration is %f sec\n", (double)(clock() - start)/CLOCKS_PER_SEC);
}

File: multithread_python/prime.py
import time
import sys
from threading import Thread

def run():
	STEP = 10000

	limit = 10000
	if len(sys.argv) == 2: limit = int(sys.argv[1])

	start = time.clock()
	number = 2
	count = 0
	while True:
		primes = {}

		t0 = Thread(target=worker, args=(primes,number+0,number+STEP,number))
		t0.start()
		t1 = Thread(target=worker, args=(primes,number+1,number+STEP,number))
		t1.start()
		t2 = Thread(target=worker, args=(primes,number+2,number+STEP,number))
		t2.start()
		t3 = Thread(target=worker, args=(primes,number+3,number+STEP,number))
		t3.start()
		t0.join()
		t1.join()
		t2.join()
		t3.join()

		for s in range(0, STEP, 1):
			if not primes[s]:
				continue
			count += 1
			if count == limit:
				print(str(limit) + "th prime is " + str(number+s))
				break
		else:
			number += STEP
			continue
		break
	print("[mt.python] duration is " + str(time.clock() - start) + " sec")

def worker(box, start, end, offset):
	for n in range(start, end, 4):
		box[n - offset] = is_prime(n)

def is_prime(number):
	for n in range(2, int(number/2)+1):
		if number % n == 0:
			return False
	return True

run()

